// @ng-bootstrap/ng-bootstrap/esm5/buttons/label.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var NgbButtonLabel = /** @class */ (function () {
    function NgbButtonLabel() {
    }
NgbButtonLabel.ngDirectiveDef = ɵngcc0.ɵɵdefineDirective({ type: NgbButtonLabel, selectors: [["", "ngbButtonLabel", ""]], factory: function NgbButtonLabel_Factory(t) { return new (t || NgbButtonLabel)(); }, hostBindings: function NgbButtonLabel_HostBindings(rf, ctx, elIndex) { if (rf & 1) {
        ɵngcc0.ɵɵallocHostVars(4);
        ɵngcc0.ɵɵstyling();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("btn", true);
        ɵngcc0.ɵɵclassProp("active", ctx.active);
        ɵngcc0.ɵɵclassProp("disabled", ctx.disabled);
        ɵngcc0.ɵɵclassProp("focus", ctx.focused);
        ɵngcc0.ɵɵstylingApply();
    } } });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(NgbButtonLabel, [{
        type: Directive,
        args: [{
                selector: '[ngbButtonLabel]',
                host: { '[class.btn]': 'true', '[class.active]': 'active', '[class.disabled]': 'disabled', '[class.focus]': 'focused' }
            }]
    }], function () { return []; }, null);
    return NgbButtonLabel;
}());
export { NgbButtonLabel };
if (false) {
    /** @type {?} */
    NgbButtonLabel.prototype.active;
    /** @type {?} */
    NgbButtonLabel.prototype.disabled;
    /** @type {?} */
    NgbButtonLabel.prototype.focused;
}
