// ngx-spinner/esm5/lib/ngx-spinner.enum.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/ngx-spinner.enum.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var LOADERS = {
    'ball-8bits': 16,
    'ball-atom': 4,
    'ball-beat': 3,
    'ball-circus': 5,
    'ball-climbing-dot': 4,
    'ball-clip-rotate': 1,
    'ball-clip-rotate-multiple': 2,
    'ball-clip-rotate-pulse': 2,
    'ball-elastic-dots': 5,
    'ball-fall': 3,
    'ball-fussion': 4,
    'ball-grid-beat': 9,
    'ball-grid-pulse': 9,
    'ball-newton-cradle': 4,
    'ball-pulse': 3,
    'ball-pulse-rise': 5,
    'ball-pulse-sync': 3,
    'ball-rotate': 1,
    'ball-running-dots': 5,
    'ball-scale': 1,
    'ball-scale-multiple': 3,
    'ball-scale-pulse': 2,
    'ball-scale-ripple': 1,
    'ball-scale-ripple-multiple': 3,
    'ball-spin': 8,
    'ball-spin-clockwise': 8,
    'ball-spin-clockwise-fade': 8,
    'ball-spin-clockwise-fade-rotating': 8,
    'ball-spin-fade': 8,
    'ball-spin-fade-rotating': 8,
    'ball-spin-rotate': 2,
    'ball-square-clockwise-spin': 8,
    'ball-square-spin': 8,
    'ball-triangle-path': 3,
    'ball-zig-zag': 2,
    'ball-zig-zag-deflect': 2,
    'cog': 1,
    'cube-transition': 2,
    'fire': 3,
    'line-scale': 5,
    'line-scale-party': 5,
    'line-scale-pulse-out': 5,
    'line-scale-pulse-out-rapid': 5,
    'line-spin-clockwise-fade': 8,
    'line-spin-clockwise-fade-rotating': 8,
    'line-spin-fade': 8,
    'line-spin-fade-rotating': 8,
    'pacman': 6,
    'square-jelly-box': 2,
    'square-loader': 1,
    'square-spin': 1,
    'timer': 1,
    'triangle-skew-spin': 1
};
/** @type {?} */
export var DEFAULTS = {
    BD_COLOR: 'rgba(51,51,51,0.8)',
    SPINNER_COLOR: '#fff',
    SPINNER_TYPE: 'ball-scale-multiple',
    Z_INDEX: 99999,
};
/** @type {?} */
export var PRIMARY_SPINNER = 'primary';
/**
 * @record
 */
export function Spinner() { }
if (false) {
    /** @type {?|undefined} */
    Spinner.prototype.bdColor;
    /** @type {?|undefined} */
    Spinner.prototype.size;
    /** @type {?|undefined} */
    Spinner.prototype.color;
    /** @type {?|undefined} */
    Spinner.prototype.type;
    /** @type {?|undefined} */
    Spinner.prototype.fullScreen;
    /** @type {?|undefined} */
    Spinner.prototype.zIndex;
}
var NgxSpinner = /** @class */ (function () {
    function NgxSpinner(init) {
        Object.assign(this, init);
    }
    return NgxSpinner;
}());
export { NgxSpinner };
if (false) {
    /** @type {?} */
    NgxSpinner.prototype.name;
    /** @type {?} */
    NgxSpinner.prototype.bdColor;
    /** @type {?} */
    NgxSpinner.prototype.size;
    /** @type {?} */
    NgxSpinner.prototype.color;
    /** @type {?} */
    NgxSpinner.prototype.type;
    /** @type {?} */
    NgxSpinner.prototype.class;
    /** @type {?} */
    NgxSpinner.prototype.divCount;
    /** @type {?} */
    NgxSpinner.prototype.divArray;
    /** @type {?} */
    NgxSpinner.prototype.fullScreen;
    /** @type {?} */
    NgxSpinner.prototype.show;
    /** @type {?} */
    NgxSpinner.prototype.zIndex;
}
