// @ng-bootstrap/ng-bootstrap/esm2015/carousel/carousel-config.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * A configuration service for the [NgbCarousel](#/components/carousel/api#NgbCarousel) component.
 *
 * You can inject this service, typically in your root component, and customize its properties
 * to provide default values for all carousels used in the application.
 */
import * as ɵngcc0 from '@angular/core';
export class NgbCarouselConfig {
    constructor() {
        this.interval = 5000;
        this.wrap = true;
        this.keyboard = true;
        this.pauseOnHover = true;
        this.showNavigationArrows = true;
        this.showNavigationIndicators = true;
    }
}
NgbCarouselConfig.ngInjectableDef = ɵngcc0.ɵɵdefineInjectable({ token: NgbCarouselConfig, factory: function NgbCarouselConfig_Factory(t) { return new (t || NgbCarouselConfig)(); }, providedIn: 'root' });
/*@__PURE__*/ ɵngcc0.ɵsetClassMetadata(NgbCarouselConfig, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, { constructor: [], interval: [], wrap: [], keyboard: [], pauseOnHover: [], showNavigationArrows: [], showNavigationIndicators: [] });
/** @nocollapse */ NgbCarouselConfig.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function NgbCarouselConfig_Factory() { return new NgbCarouselConfig(); }, token: NgbCarouselConfig, providedIn: "root" });
if (false) {
    /** @type {?} */
    NgbCarouselConfig.prototype.interval;
    /** @type {?} */
    NgbCarouselConfig.prototype.wrap;
    /** @type {?} */
    NgbCarouselConfig.prototype.keyboard;
    /** @type {?} */
    NgbCarouselConfig.prototype.pauseOnHover;
    /** @type {?} */
    NgbCarouselConfig.prototype.showNavigationArrows;
    /** @type {?} */
    NgbCarouselConfig.prototype.showNavigationIndicators;
}
